### PR TITLE
Add memberships endpoints

### LINKED ISSUE #876: Add /key.get and /key.get_accounts

### LINKED ISSUE #881: Refactor the way memberships are handled

### PR DESCRIPTION
**Pull request by mederic-p**
Issue/Task Number: 876
Closes #876, closes #881 

# Overview

This PR adds a few new endpoints to retrieve memberships on accounts, admins and keys.
It also deprecates the `account.get_members` that was returning a list of admins in favor of the new `account.get_admin_user_memberships` that returns a list of memberships.

# Changes

- Deprecate `account.get_members`
- Remove `account.get_keys` (was not released yet so we can safely remove it) and replace it with `account.get_key_memberships`
- Add `account.get_admin_user_memberships`
- Add `account.get_key_memberships`
- Add `key.get_account_memberships`
- Add `admin.get_account_memberships`
- Add `key.get`

# Implementation Details

It has been decided to return a new `membership` structure that contains relation between `key`, `admin` and `account` for all related endpoints.
The `account.get_members` still work and returns a list of users with extra attributes but is now deprecated.
All deprecated function related to this endpoint were extracted moved to new files containing `Deprecated` in their name to facilitate future removal.

# Usage

Open API spec has been updated, so the easiest way to test this endpoint is to try them directly in swagger UI.

# Impact

No DB changes.
New endpoints
Deprecate `account.get_members`

### CODE REVIEW

File: apps/admin_api/lib/admin_api/v1/controllers/account_membership_controller.ex [hunk 1]
```diff
@@ -28,13 +29,62 @@ defmodule AdminAPI.V1.AccountMembershipController do
   alias EWalletDB.{Account, Key, Membership, Role, User}
 
   @doc """
-  Lists the users that are assigned to the given account.
+  Lists the memberships for the given admin.
   """
-  def all_users_for_account(conn, attrs) do
+  def all_account_memberships_for_admin(conn, %{"id" => admin_id} = attrs) do
+    with %User{} = admin <-
+           User.get_admin(admin_id, preload: [memberships: [:user, :role]]) ||
```
Comment (T-Dnzt) [inline apps/admin_api/lib/admin_api/v1/controllers/account_membership_controller.ex]: I don't think we need to preload anything here anymore 🤔

File: apps/admin_api/lib/admin_api/v1/controllers/account_membership_controller.ex [hunk 2]
```diff
@@ -28,13 +29,62 @@ defmodule AdminAPI.V1.AccountMembershipController do
...
+  end
+
+  def all_account_memberships_for_admin(conn, _), do: handle_error(conn, :missing_id)
+
+  @doc """
+  Lists the memberships for the given key.
+  """
+  def all_account_memberships_for_key(conn, %{"id" => key_id} = attrs) do
+    with %Key{} = key <-
+           Key.get(key_id, preload: [memberships: [:key, :role]]) ||
```
Comment (T-Dnzt) [inline apps/admin_api/lib/admin_api/v1/controllers/account_membership_controller.ex]: Same here, probably.

File: apps/admin_api/lib/admin_api/v1/controllers/account_membership_controller.ex [hunk 3]
```diff
@@ -28,13 +29,62 @@ defmodule AdminAPI.V1.AccountMembershipController do
...
-  Lists the users that are assigned to the given account.
+  Lists the memberships for the given admin.
   """
-  def all_users_for_account(conn, attrs) do
+  def all_account_memberships_for_admin(conn, %{"id" => admin_id} = attrs) do
+    with %User{} = admin <-
+           User.get_admin(admin_id, preload: [memberships: [:user, :role]]) ||
+             {:error, :unauthorized},
+         {:ok, _} <- authorize(:get, conn.assigns, admin),
+         {:ok, %{query: query}} <- authorize(:all, conn.assigns, nil),
```
Comment (T-Dnzt) [inline apps/admin_api/lib/admin_api/v1/controllers/account_membership_controller.ex]: The `query` returned can be `nil` if the actor has no permissions. Need to check with

```
true <- !is_nil(query) || {:error, :unauthorized} do
```

Can you add a test as well (with global role `none`)?

File: apps/admin_api/lib/admin_api/v1/controllers/account_membership_controller.ex [hunk 4]
```diff
@@ -28,13 +29,62 @@ defmodule AdminAPI.V1.AccountMembershipController do
...
+
+  @doc """
+  Lists the memberships for the given key.
+  """
+  def all_account_memberships_for_key(conn, %{"id" => key_id} = attrs) do
+    with %Key{} = key <-
+           Key.get(key_id, preload: [memberships: [:key, :role]]) ||
+             {:error, :unauthorized},
+         {:ok, _} <- authorize(:get, conn.assigns, key),
+         {:ok, %{query: query}} <- authorize(:all, conn.assigns, nil),
```
Comment (T-Dnzt) [inline apps/admin_api/lib/admin_api/v1/controllers/account_membership_controller.ex]: Same for checking if `query = nil`.

File: apps/admin_api/lib/admin_api/v1/controllers/account_membership_deprecated_controller.ex [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+  @doc """
+  Lists the admins that are assigned to the given account.
+  """
+  def all_admin_for_account(conn, %{"id" => account_id} = attrs) do
+    with %Account{} = account <-
+           Account.get(account_id, preload: [memberships: [:user, :role]]) ||
+             {:error, :unauthorized},
+         {:ok, _} <- authorize(:get, conn.assigns, account),
+         {:ok, %{query: query}} <- authorize(:all, conn.assigns, nil),
```
Comment (T-Dnzt) [inline apps/admin_api/lib/admin_api/v1/controllers/account_membership_deprecated_controller.ex]: `true <- !is_nil(query) || {:error, :unauthorized},`

File: apps/ewallet_db/lib/ewallet_db/membership.ex [hunk 6]
```diff
@@ -98,43 +98,50 @@ defmodule EWalletDB.Membership do
...
+  @spec query_all_by_key(EWalletDB.Key.t(), any(), any()) :: Ecto.Query.t()
+  def query_all_by_key(key, query \\ Membership, preload \\ []) do
+    query
+    |> where([m], m.key_uuid == ^key.uuid)
+    |> preload([m], ^preload)
   end
 
-  @spec all_by_account(EWalletDB.Account.t(), any(), any()) :: Ecto.Query.t()
-  def all_by_account(%Account{} = account, query \\ Membership, preload \\ []) do
+  @spec query_all_by_account(EWalletDB.Account.t(), any(), any()) :: Ecto.Query.t()
```
Comment (unnawut) [inline apps/ewallet_db/lib/ewallet_db/membership.ex]: 🙏 for refactoring the names

File: apps/ewallet_db/lib/ewallet_db/user.ex [hunk 7]
```diff
@@ -355,14 +355,13 @@ defmodule EWalletDB.User do
...
   """
-  @spec get_admin(String.t()) :: %User{} | nil
-  @spec get_admin(String.t(), Ecto.Queryable.t()) :: %User{} | nil
-  def get_admin(id, queryable \\ User)
+  @spec get_admin(String.t(), keyword()) :: %__MODULE__{} | nil
+  def get_admin(id, opts \\ [:wallets])
 
-  def get_admin(id, queryable) when is_external_id(id) do
-    queryable
+  def get_admin(id, opts) when is_external_id(id) do
```
Comment (unnawut) [inline apps/ewallet_db/lib/ewallet_db/user.ex]: 🙏 for refactoring for consistent args
